ivadomed/loader/mri2d_segmentation_dataset.py

```diff
@@ -196,81 +196,92 @@ def __getitem__(self, index):
-
-            data_dict = {
-                'input': torch.zeros(stack_input.shape[0], shape_x, shape_y),
-                'gt': torch.zeros(stack_gt.shape[0], shape_x, shape_y) if stack_gt is not None else None,
-                'roi': torch.zeros(stack_roi.shape[0], shape_x, shape_y) if stack_roi is not None else None,
-                'input_metadata': metadata_input,
-                'gt_metadata': metadata_gt,
-                'roi_metadata': metadata_roi
-            }
-
-            for _ in range(len(stack_input)):
-                data_dict['input'] = stack_input[:,
+                metadata['coord'] = [coord["x_min"], coord["x_max"],
+                                     coord["y_min"], coord["y_max"]]
+
+            # Extract image and gt patches
+            stack_roi = None
+            metadata_roi = None
+
+            patch_input = np.asarray(seg_pair_slice["input"])[:,
+                                     coord['x_min']:coord['x_max'],
+                                     coord['y_min']:coord['y_max']]
+
+            if seg_pair_slice["gt"]:
+                patch_gt = np.asarray(seg_pair_slice["gt"])[:,
                                       coord['x_min']:coord['x_max'],
                                       coord['y_min']:coord['y_max']]
+            else:
+                patch_gt = []
+
+            # Run transforms on image patches
+            stack_input, metadata_input = self.transform(sample=list(patch_input),
+                                                         metadata=metadata_input,
+                                                         data_type="im")
+
+            # Update metadata_gt with metadata_input
+            metadata_gt = imed_loader_utils.update_metadata(metadata_input, metadata_gt)
+
+            if self.task == "segmentation":
+                # Run transforms on gt patches
+                stack_gt, metadata_gt = self.transform(sample=list(patch_gt),
+                                                       metadata=metadata_gt,
+                                                       data_type="gt")
+                # Make sure stack_gt is binarized
+                if stack_gt is not None and not self.soft_gt:
+                    stack_gt = imed_postpro.threshold_predictions(stack_gt, thr=0.5).astype(np.uint8)
+            else:
+                # Force no transformation on labels for classification task
+                # stack_gt is a tensor of size 1x1, values: 0 or 1
+                # "expand(1)" is necessary to be compatible with segmentation convention: n_labelxhxwxd
+                stack_gt = torch.from_numpy(seg_pair_slice["gt"][0]).expand(1)
 
-            if stack_gt is not None:
-                for _ in range(len(stack_gt)):
-                    data_dict['gt'] = stack_gt[:,
-                                       coord['x_min']:coord['x_max'],
-                                       coord['y_min']:coord['y_max']]
+        else:
```

Comment:
There seems to be a lot of code duplication here between `if self.is_2d_patch` and the `else`. Could we simply keep the code under the first part of the if and simply set 
```python
if not self.is_2d_patch:
    coord = {}
    coord['x_min'], coord['x_max'], coord['y_min'], coord['y_max'] = 0, seg_pair_input["input"][0].shape[0], 0, seg_pair_input["input"][0].shape[1]
``` 
(something like that) I might have missed some technicalities why this is not possible, so please ignore if it's too complicated.

Reply:
I had tried to minimize the duplicated code but could not find a good way. I think your suggestion could work well, thank you!
I'll try it, the only issue I might have is to test the ROI correctly. Currently ROI only applies to entire slice and not patches and I'm not very familiar with the feature. Is the [Cascaded architecture](https://ivadomed.org/tutorials/cascaded_architecture.html#cascaded-architecture) tutorial using this feature? (by cropping around the segmented spinal cord?)

Reply:
Thank you @andreanne-lemay!
I was able to remove the code duplication following your advice, it looks a lot better!

I answered my own question: the "cascaded" tutorial does not use the ROI feature. I ended up testing it with `ROICrop` and `suffix` in `roi_params` which I think is the right thing to do.

I tested several scenarios (Microscopy 2D with patches, MRI 2D with and without patches, and MRI with ROICrop.
Everything looks good to me and the tests are passing.

Would you be available to take a look again to make sure I did not missed something before I merge?

As for the 3D loader, I will open a separate issue as I don't have time or resources to take of it at the moment.
Thank you again!

Reply:
Thanks for looking into it and sounds good for the separate 3D issue! I'm looking through it again now, sorry for the delay!